cpp/src/io/parquet/parquet_reader.cu

```diff
@@ -0,0 +1,826 @@
+    }
+    if (fd) {
+      close(fd);
+    }
+  }
+
+  const uint8_t *data() const { return static_cast<uint8_t *>(mapped_data); }
+  size_t size() const { return mapped_size; }
+
+ private:
+  void *mapped_data = nullptr;
+  size_t mapped_size = 0;
+  int fd = 0;
+};
+
+/**
+ * @brief Function that translates Parquet datatype to GDF dtype
+ **/
+constexpr std::pair<gdf_dtype, gdf_dtype_extra_info> to_dtype(
+    parquet::Type physical, parquet::ConvertedType logical) {
+  // Logical type used for actual data interpretation; the legacy converted type
+  // is superceded by 'logical' type whenever available.
+  switch (logical) {
+    case parquet::UINT_8:
+    case parquet::INT_8:
+      return std::make_pair(GDF_INT8, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::UINT_16:
+    case parquet::INT_16:
+      return std::make_pair(GDF_INT16, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::DATE:
+      return std::make_pair(GDF_DATE32, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::TIMESTAMP_MILLIS:
+      return std::make_pair(GDF_DATE64, gdf_dtype_extra_info{TIME_UNIT_ms});
+    case parquet::TIMESTAMP_MICROS:
+      return std::make_pair(GDF_DATE64, gdf_dtype_extra_info{TIME_UNIT_us});
+    default:
+      break;
+  }
+
+  // Physical storage type supported by Parquet; controls the on-disk storage
+  // format in combination with the encoding type.
+  switch (physical) {
+    case parquet::BOOLEAN:
+      return std::make_pair(GDF_INT8, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::INT32:
+      return std::make_pair(GDF_INT32, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::INT64:
+      return std::make_pair(GDF_INT64, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::FLOAT:
+      return std::make_pair(GDF_FLOAT32, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::DOUBLE:
+      return std::make_pair(GDF_FLOAT64, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::BYTE_ARRAY:
+    case parquet::FIXED_LEN_BYTE_ARRAY:
+      // Can be mapped to GDF_CATEGORY (32-bit hash) or GDF_STRING (nvstring)
+      return std::make_pair(GDF_CATEGORY, gdf_dtype_extra_info{TIME_UNIT_NONE});
+    case parquet::INT96:
+      // deprecated, only used by legacy implementations
+    default:
+      break;
+  }
+
+  return std::make_pair(GDF_invalid, gdf_dtype_extra_info{TIME_UNIT_NONE});
+}
+
+/**
+ * @brief Helper that returns the required the number of bits to store a value
+ **/
+template <typename T = uint8_t>
+T required_bits(uint32_t max_level) {
+  return static_cast<T>(parquet::CPReader::NumRequiredBits(max_level));
+}
+
+/**
+ * @brief A helper class that wraps a gdf_column and any associated memory.
+ *
+ * This abstraction initializes and manages a gdf_column (fields and memory)
+ * while still allowing direct access. Memory is automatically deallocated
+ * unless ownership is transferred via releasing and assigning the raw pointer.
+ **/
+class gdf_column_wrapper {
```

Comment:
Don't we have this functionality somewhere in cudf test utilities?

Reply:
There's a similar class called column_wrapper, but that's geared towards unit tests, and initialized with host memory.

What we need for the cuIO readers is a lightweight pointer wrapper, for just de-allocation in case of exceptions. The underlying gdf_column will be returned to the caller if successful and only the wrapper is discarded.

We could modify that class for our needs, but we would end up not using most of the existing code there (like host vector initialization, etc.)...

Reply:
OK. I think we have plans to replace `gdf_column` with `cudf::column` which will be a proper C++ class with initialization, etc.  Once we have that, revisit.